core/src/solve.jl

```diff
@@ -256,21 +301,88 @@ function formulate!(du::AbstractVector, basin::Basin, u::AbstractVector, t::Real
         drainage = basin.drainage[i]
         infiltration = reduction_factor * basin.infiltration[i]
 
-        du[i] += precipitation - evaporation + drainage - infiltration
+        du.storage[i] += precipitation - evaporation + drainage - infiltration
     end
     return nothing
 end
 
+function get_error(pid_control::PidControl, p::Parameters)::Vector{Float64}
+    (; basin) = p
+    (; node_id, listen_node_id) = pid_control
+
+    error = zero(pid_control.derivative)
+
+    for (i, id) in enumerate(node_id)
+        listened_node_id = listen_node_id[i]
+        has_index, listened_node_idx = id_index(basin.node_id, listened_node_id)
+        @assert has_index "Listen node $listened_node_id is not a basin."
+        target_level = basin.target_level[listened_node_idx]
+        error[i] = target_level - basin.current_level[listened_node_idx]
+    end
+
+    return error
+end
+
+function continuous_control!(
+    du::AbstractVector,
+    pid_control::PidControl,
+    p::Parameters,
+    integral_::Vector{Float64},
+    dstorage_old::Vector{Float64},
+)::Nothing
+    # TODO: Also support being able to control weir
+    # TODO: also support time varying target levels
+    # TODO: Also support pump flow rate to be between prescribed bounds
+    (; connectivity, pump, basin) = p
+    (; graph_control) = connectivity
+    (; node_id, proportional, derivative, integral, listen_node_id) = pid_control
+
+    error_ = get_error(pid_control, p)
+
+    for (i, id) in enumerate(node_id)
+        error = error_[i]
+        du.integral[i] += error
+
+        listened_node_id = listen_node_id[i]
+        _, listened_node_idx = id_index(basin.node_id, listened_node_id)
+
+        flow_rate = 0.0
+
+        if !isnan(proportional[i])
+            flow_rate += proportional[i] * error
+        end
+
+        if !isnan(derivative[i])
+            # dlevel/dstorage = 1/area
+            area = basin.current_area[listened_node_idx]
+
+            error_deriv = -dstorage_old[listened_node_idx] / area
+            flow_rate += derivative[i] * error_deriv
+        end
+
+        if !isnan(integral[i])
+            # coefficient * current value of integral
+            flow_rate += integral[i] * integral_[i]
+        end
+
+        flow_rate = max(flow_rate, 0.0)
+
+        controlled_node_id = only(outneighbors(graph_control, id))
+        controlled_node_idx = findfirst(pump.node_id .== controlled_node_id) # TODO: support the use of id_index
```

Comment:
Best to put comments like this on the line above, otherwise the line becomes quite long.